File: mobile-app/lib/features/notifications/notifications_inbox.dart
/// Pure-dart in-memory Inbox für PUSH_NOTIFICATION-Events vom sync-server.
/// Keine Flutter-Imports — testbar ohne widget-tester.
///
/// Items sind die rohen Maps aus sync-server/lib/notify_dispatch.js
/// (Felder: id, type, title, body, projectId, priority, createdAt).
/// Reihenfolge in [list]: neueste zuerst. Re-push gleicher id bringt das
/// Item nach vorne ohne Read-Status zu resetten.
class NotificationsInbox {
  NotificationsInbox({this.max = 50});

  final int max;
  final List<Map<String, dynamic>> _items = [];
  final Set<String> _read = <String>{};

  static const _required = ['id', 'type', 'title', 'body', 'createdAt'];

  bool _isValid(Object? raw) {
    if (raw is! Map) return false;
    for (final k in _required) {
      final v = raw[k];
      if (v is! String || v.isEmpty) return false;
    }
    return true;
  }

  bool push(Object? raw) {
    if (!_isValid(raw)) return false;
    final item = Map<String, dynamic>.from(raw as Map);
    final id = item['id'] as String;
    _items.removeWhere((x) => x['id'] == id);
    _items.insert(0, item);
    while (_items.length > max) {
      final dropped = _items.removeLast();
      _read.remove(dropped['id']);
    }
    return true;
  }

  List<Map<String, dynamic>> get list => List.unmodifiable(_items);

  int get unreadCount =>
      _items.where((x) => !_read.contains(x['id'])).length;

  bool isRead(String id) => _read.contains(id);

  void markRead(String id) {
    if (_items.any((x) => x['id'] == id)) _read.add(id);
  }

  void markAllRead() {
    for (final x in _items) {
      _read.add(x['id'] as String);
    }
  }

  void drop(String id) {
    _items.removeWhere((x) => x['id'] == id);
    _read.remove(id);
  }

  void clear() {
    _items.clear();
    _read.clear();
  }
}

File: desktop-app/notifications_inbox.js
"use strict";

/**
 * Pure helper: in-memory Inbox für PUSH_NOTIFICATION-Events vom sync-server.
 * Keine I/O, keine DOM-Zugriffe — testbar mit node:test.
 *
 * Public API: createInbox({ max }) → { push, list, unreadCount, markRead,
 *   markAllRead, drop, clear }
 *
 * Items sind die vom server gelieferten notification-Objekte aus
 * lib/notify_dispatch.js (id, type, title, body, projectId, priority,
 * createdAt). Reihenfolge in list(): neueste zuerst. Cap (default 50) wirft
 * älteste raus. Re-push gleicher id ersetzt das Item ohne Status-Reset.
 */

const NOTIF_INBOX_DEFAULT_MAX = 50;
const REQUIRED = ["id", "type", "title", "body", "createdAt"];

function isValid(item) {
  if (!item || typeof item !== "object") return false;
  for (const k of REQUIRED) if (typeof item[k] !== "string" || !item[k]) return false;
  return true;
}

function createInbox(opts) {
  const max = (opts && opts.max) || NOTIF_INBOX_DEFAULT_MAX;
  const items = [];
  const read = new Set();

  function push(item) {
    if (!isValid(item)) return false;
    const idx = items.findIndex(x => x.id === item.id);
    if (idx >= 0) { items.splice(idx, 1); }
    items.unshift({ ...item });
    while (items.length > max) {
      const dropped = items.pop();
      read.delete(dropped.id);
    }
    return true;
  }

  function list() { return items.slice(); }
  function unreadCount() {
    return items.reduce((c, x) => c + (read.has(x.id) ? 0 : 1), 0);
  }
  function markRead(id) { if (items.some(x => x.id === id)) read.add(id); }
  function markAllRead() { for (const x of items) read.add(x.id); }
  function drop(id) {
    const i = items.findIndex(x => x.id === id);
    if (i >= 0) { items.splice(i, 1); read.delete(id); }
  }
  function clear() { items.length = 0; read.clear(); }

  return { push, list, unreadCount, markRead, markAllRead, drop, clear };
}

if (typeof module !== "undefined") module.exports = { createInbox };
if (typeof window !== "undefined") window.createNotificationsInbox = createInbox;
